connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala

```diff
@@ -0,0 +1,140 @@
+import io.unitycatalog.client.model.TableType
+
+import java.net.URI
+import java.util
+
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException
+import org.apache.spark.sql.catalyst.catalog.{CatalogStorageFormat, CatalogTable, CatalogTableType, CatalogUtils}
+import org.apache.spark.sql.connector.catalog._
+import org.apache.spark.sql.connector.expressions.Transform
+import org.apache.spark.sql.types.{DataType, StructField, StructType}
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+import scala.collection.JavaConverters._
+
+/**
+ * A Spark catalog plugin to get/manage tables in Unity Catalog.
+ */
+class UCSingleCatalog extends TableCatalog {
+
+  private var deltaCatalog: TableCatalog = null
+
+  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
+    val proxy = new UCProxy()
+    proxy.initialize(name, options)
+    deltaCatalog = Class.forName("org.apache.spark.sql.delta.catalog.DeltaCatalog")
+      .getDeclaredConstructor().newInstance().asInstanceOf[TableCatalog]
+    deltaCatalog.asInstanceOf[DelegatingCatalogExtension].setDelegateCatalog(proxy)
+  }
+
+  override def name(): String = deltaCatalog.name()
+
+  override def listTables(namespace: Array[String]): Array[Identifier] = ???
+
+  override def loadTable(ident: Identifier): Table = deltaCatalog.loadTable(ident)
+
+  override def tableExists(ident: Identifier): Boolean = {
+    deltaCatalog.tableExists(ident)
+  }
+
+  override def createTable(
+      ident: Identifier,
+      columns: Array[Column],
+      partitions: Array[Transform],
+      properties: util.Map[String, String]): Table = {
+    deltaCatalog.createTable(ident, columns, partitions, properties)
+  }
+
+  override def createTable(ident: Identifier, schema: StructType, partitions: Array[Transform], properties: util.Map[String, String]): Table = {
+    throw new AssertionError("deprecated `createTable` should not be called")
+  }
+  override def alterTable(ident: Identifier, changes: TableChange*): Table = ???
+
+  override def dropTable(ident: Identifier): Boolean = ???
+
+  override def renameTable(oldIdent: Identifier, newIdent: Identifier): Unit = ???
+}
+
+// An internal proxy to talk to the UC client.
+private class UCProxy extends TableCatalog {
+  private[this] var name: String = null
+  private[this] var tablesApi: TablesApi = null
+
+  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
+    this.name = name
+    val urlStr = options.get("url")
+    if (urlStr == null) {
+      throw new IllegalArgumentException("url must be specified for Unity Catalog")
+    }
+    val url = new URI(urlStr)
+    var client = new ApiClient()
+      .setHost(url.getHost)
+      .setPort(url.getPort)
+      .setScheme(url.getScheme)
+    val token = options.get("token")
+    if (token != null && token.nonEmpty) {
+      client = client.setRequestInterceptor(request => {
```

Comment:
`setRequestInterceptor { request => ` would be more Scala-esque (and avoid this extra `{`)